tests/runtimes/android/android_runtime_tests.py

```diff
@@ -0,0 +1,271 @@
+        Tns.platform_add_android(APP_NAME, framework_path=Android.FRAMEWORK_PATH)
+
+        source = os.path.join(TEST_RUN_HOME, 'assets', 'runtime', 'android', 'files', 'android-runtime-1160',
+                              'testActivity.android.js')
+        target = os.path.join(TEST_RUN_HOME, APP_NAME, 'app')
+        File.copy(src=source, target=target)
+
+        Tns.build_android(os.path.join(TEST_RUN_HOME, APP_NAME))
+        activity_class_path = os.path.join(TEST_RUN_HOME, APP_NAME, "platforms", "android", "app", "src", "main",
+                                           "java", "com",
+                                           "test")
+
+        if File.exists(os.path.join(activity_class_path, "Activity.java")):
+            assert True
+        else:
+            assert False, "Fail: Custom activity class is NOT generated in {0} !".format(activity_class_path)
+
+    def test_390_code_cache_option(self):
+        """
+        CodeCache option is broken since Android Runtime 4.1.0
+
+        https://github.com/NativeScript/android-runtime/issues/1235
+        """
+        Folder.clean(os.path.join(TEST_RUN_HOME, APP_NAME))
+        Tns.create(app_name=APP_NAME, template=Template.HELLO_WORLD_JS.local_package, update=True)
+
+        source_js = os.path.join(TEST_RUN_HOME, 'assets', 'runtime', 'android', 'files', 'android-runtime-1235',
+                                 'package.json')
+        target_js = os.path.join(TEST_RUN_HOME, APP_NAME, 'app', 'package.json')
+        File.copy(src=source_js, target=target_js)
+
+        Tns.platform_add_android(APP_NAME, framework_path=Android.FRAMEWORK_PATH)
+
+        # `tns run android` and wait until app is deployed
+        log = Tns.run_android(APP_NAME, device=self.emulator.id, wait=False, verify=False)
+
+        strings = ['Project successfully built',
+                   'Successfully installed on device with identifier', self.emulator.id,
+                   'Successfully synced application']
+        test_result = Wait.until(lambda: all(string in File.read(log.log_file) for string in strings), timeout=300,
+                                 period=5)
+
+        assert test_result, 'Application not build successfully!'
+
+        code_cache_files = ['app.js.cache', 'main-page.js.cache', 'main-view-model.js.cache']
+        json = App.get_package_json(app_name=APP_NAME)
+        app_id = json['nativescript']['id']
+
+        # Check that for each .js file, there's a corresponding .js.cache file created on the device
+        for code_cache_file in code_cache_files:
+            error_message = '{0} file is not found on {1}'.format(code_cache_file, self.emulator.id)
+            assert Adb.file_exists(device_id=self.emulator.id, package_id=app_id,
+                                   file_name='app/{0}'.format(code_cache_file)), error_message
+
+        # Verify app looks correct inside emulator
+        Device.screen_match(self.emulator,
```

Comment:
One of the main purposes of the new repo is making tests independent on infrastructure.
They should be able to run on other emulators as well and do not depend on how we draw layouts.
Bringing screenshots will also increase repo size and will make git clone slower.
Can you assert that text is visible instead of comparing screenshots?

Reply:
The cases we use screen match are very rare. We have 2 images for all android runtime tests. We use them to check cases when the app could have very strange behavior for example catching native exception. We are trying to avoid them.